blog/content/post/2024-08-19-secureboot-freebsd.md

```diff
@@ -0,0 +1,397 @@
+directly.
+
+## Creating a Secure Boot - ready FreeBSD EFI executable
+
+Please note that this guide assumes that you have a working FreeBSD
+installation that you wish to modify for Secure Boot support. Installation of
+the system will not be covered here.
+
+### Determine the loader of choice
+
+There are currently several flavors of the `loader.efi` available. The original
+one was written in
+[FORTH](https://hackaday.com/2017/01/27/forth-the-hackers-language/).
+The 14.1 release defaults to a more modern Lua one, the final option being
+`loader_simp` - a simplified C implementation. This guide will only provide
+instructions on using the legacy FORTH loader and the current default.
+
+### Copy the boot filesystem
+
+```bash
+mkdir ~/bootfs
+mkdir ~/bootfs/boot
+cd ~/bootfs/boot
+cp -r /boot/kernel .
+cp -r /boot/defaults .
+cp /boot/loader.conf .
+cp /boot/*.rc .
+cp /boot/device.hints .
+cp /boot/loader.help* .
+```
+
+#### Lua loader
+
+```bash
+mkdir lua
+cp -r /boot/lua ./lua
+```
+
+#### FORTH loader
+
+```bash
+cp /boot/*.4th .
+```
+
+### Copy the fstab
+
+```bash
+mkdir ~/bootfs/etc
+cp /etc/fstab ~/bootfs/etc/fstab
+```
+
+### Create an image of the filesystem
+
+```bash
+cd ~/
+makefs bootfs.img bootfs
+```
+
+Determine the size of the boot filesystem. This will determine how much memory
+to reserve in the loader:
+
+```bash
+ls -l bootfs.img | awk '{print $5}'
+```
+
+Add a safety factor of a couple hundred bytes (~512) to this number, and record
+it.
+
+### Build the Loader With Extra Space
+
+This step requires you to have the source code for your system. You might have
+checked the `src` component during OS installation, then the `/usr/src/`
+directory should be already populated. If not, clone the source code from
+GitHub, substituting `14.1` with the release you are currently using:
+
+```bash
+git clone -b releng/14.1 https://git.freebsd.org/src.git /usr/src
+```
+
+Now, recall your bootfs size and substitute `${BOOTFS_SIZE_PLUS_SAFETY}` below:
+
+```bash
+cd /usr/src/stand
+make MD_IMAGE_SIZE=${BOOTFS_SIZE_PLUS_SAFETY}
+```
+
+### Embed bootfs image in the loader.efi
+
+The build results will be available under `/usr/obj`.
+
+* If you have chosen to use the legacy FORTH loader, the path to the file that
+  interests you should be similar to
+  `/usr/obj/usr/src/amd64.amd64/stand/efi/loader_4th/loader_4th.efi`.
+
+* If you have chosen the Lua loader, the path should be similar to
+  `/usr/obj/usr/src/amd64.amd64/stand/efi/loader_lua/loader_lua.efi`
+
+Copy the appropriate file and use the available embedding utility:
+
+```bash
+cp /usr/obj/${PATH_TO_LOADER}/loader.efi ~/
+/usr/src/sys/tools/embed_mfs.sh ~/loader.efi ~/bootfs.img
+```
+
+You might need to make `embed_mfs.sh` executable:
+
+```bash
+chmod +x /usr/src/sys/tools/embed_mfs.sh
+```
+
+At this point, `loader.efi` is an UEFI-bootable binary, consisting of the
+FreeBSD bootloader and kernel. The last remaining step to Secure Boot
```

Comment:
Please replace every "Secure Boot" with "UEFI Secure Boot" when you mean the second one.